src/api/utils/fetch-factory.ts

```diff
@@ -0,0 +1,127 @@
+  const contentType = response.headers.get('content-type')
+
+  const contentTypeHandlers: Record<string, (response: Response) => Promise<any>> = {
+    'application/json': response => response.json(),
+    'application/pdf': response => response.blob(),
+    'application/octet-stream': response => response.arrayBuffer(),
+    'text': response => response.text(),
+  }
+
+  const defaultHandler = (response: Response) => response.text()
+
+  const getHandler = (contentType: string | null) => {
+    if (!contentType) {
+      return defaultHandler
+    }
+
+    for (const [type, handler] of Object.entries(contentTypeHandlers)) {
+      if (contentType.includes(type)) {
+        return handler
+      }
+    }
+
+    return defaultHandler
+  }
+
+  const handler = getHandler(contentType)
+  const data = await handler(response)
+  return data as T
+}
+
+function mergeHeaders (defaultHeaders: HeadersInit = {}, requestHeaders: HeadersInit = {}): HeadersInit {
+  const headers = new Headers()
+
+  Object.entries(defaultHeaders).forEach(([key, value]) => {
+    if (Array.isArray(value)) {
+      value.forEach(v => headers.append(key, v))
+    }
+    else {
+      headers.set(key, value)
+    }
+  })
+
+  Object.entries(requestHeaders).forEach(([key, value]) => {
+    if (Array.isArray(value)) {
+      value.forEach(v => headers.append(key, v))
+    }
+    else {
+      headers.set(key, value)
+    }
+  })
+
+  return headers
+}
+
+function createCustomFetch (config: FetchConfig = {}, interceptorManager: FetchInterceptorManager = new FetchInterceptorManager()) {
+  const {
+    baseUrl = __API_URL__,
+    defaultHeaders = { 'Content-Type': 'application/json' },
+  } = config
+
+  return async <T>(url: string, options: RequestInit = {}): Promise<T> => {
+    try {
+      const interceptedOptions = await interceptorManager.applyRequestInterceptors(url, options, config)
+
+      const requestUrl = buildUrl(url, baseUrl)
+      const mergedHeaders = mergeHeaders(defaultHeaders, interceptedOptions.headers)
+
+      const requestInit: RequestInit = {
+        ...interceptedOptions,
+        headers: mergedHeaders,
+      }
+
+      const response = await fetch(requestUrl, {
+        ...requestInit,
+      })
+
+      const interceptedResponse = await interceptorManager.applyResponseInterceptors(response)
+
+      if (!interceptedResponse.ok) {
+        const errorData = await getBody(interceptedResponse.clone())
+
+        const fetchError = {
+          response: {
+            status: interceptedResponse.status,
+            statusText: interceptedResponse.statusText,
+            data: errorData,
+            headers: interceptedResponse.headers,
+          },
+          request: {
+            url: requestUrl,
+            method: requestInit.method || 'GET',
+          },
+          message: `HTTP ${interceptedResponse.status}: ${interceptedResponse.statusText}`,
+        }
+
+        throw fetchError
+      }
```

Comment:
Ici tu peux sortir la logique et faire une fonction pour la création de ton fetchError : 

```suggestion
async function createFetchError(response: Response, url: string, options: RequestInit) {
  const errorData = await getBody(response.clone())
  return {
    response: {
      status: response.status,
      statusText: response.statusText,
      data: errorData,
      headers: response.headers,
    },
    request: {
      url,
      method: options.method || 'GET',
    },
    message: `HTTP ${response.status}: ${response.statusText}`,
  }
}

      if (!interceptedResponse.ok) {
        const fetchError = await createFetchError(interceptedResponse, requestUrl, interceptedOptions)
        throw fetchError
      }
```